### DISCUSSION TITLE
Chain joins together

### DISCUSSION THREAD

**Discussion by IvesFeng** [sqlalchemy/sqlalchemy#8033] [Usage Questions] [answered]
Suppose I have 5 tables, A B C D E I know the relationship from C to E, and the relationship from A to C is determined by other factors. I have written code like the following

```
select(A.id).join_from(
    A, C, C.a_id=A.id
).join_from(
    C, D, C.id == D.c_id
).join_from(
    D, E, E.d_id== D.id
)
```
Since the C to E relationship is often used, I would like to write it as follows
```
from sqlalchemy import select, join
suffix_join = join(
    C, D, C.id == D.c_id
).join(
    D, E.d_id== D.id
)

# possible select stmt
stmt = select(A.id).join_from(
    A, C, C.a_id=A.id
)
```
But how to use "stmt" and "suffix_join" to achieve the same effect?

**Comment by zzzeek** [comment, marked as answer]
with those exact parameters, you would need to algorithmically knit those two sides together, and you would need to extract the FROM clause from the `select()` construct to do this.   The extraction part of it would involve either the use of private API members to pull out the pending join state, or getting the final assembled JOIN using the `get_final_froms()` method.  `get_final_froms()` is an expensive method that performs the whole compilation of the statement in order to determine what the FROM clause would look like.   you then would need to know that the part of your "suffix_join" which is important is the ".right" side of it, because you already have a join to C.   The example below illustrates the whole operation (noting, this is not what I would do for a real-world problem to be solved):

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import join
from sqlalchemy import select
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import join
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)

class C(Base):
    __tablename__ = 'c'

    id = Column(Integer, primary_key=True)
    a_id = Column(Integer)

class D(Base):
    __tablename__ = 'd'

    id = Column(Integer, primary_key=True)
    c_id = Column(Integer)

class E(Base):
    __tablename__ = 'e'

    id = Column(Integer, primary_key=True)
    d_id = Column(Integer)


stmt1 = select(A.id).join_from(
    A, C, C.a_id==A.id
).join_from(
    C, D, C.id == D.c_id
).join_from(
    D, E, E.d_id== D.id
)

print(stmt1)


suffix_join = join(
    C, D, C.id == D.c_id
).join(
    D, E.d_id== D.id
)

# possible select stmt
stmt2 = select(A.id).join_from(
    A, C, C.a_id==A.id
)

# this is inefficient
to_knit = stmt2.get_final_froms()[0]

stmt2 = stmt2.select_from(
    to_knit.join(suffix_join.right, C.id == D.c_id)
)

print(stmt2)
```

now how would I actually do this.   I would turn suffix_join into a callable. much easier and more efficient.

```py
def suffix_join(stmt):
    return stmt.join(D, C.id == D.c_id).join(
        D, E.d_id== D.id
    )

# possible select stmt
stmt2 = select(A.id).join_from(
    A, C, C.a_id==A.id
)

stmt2 = suffix_join(stmt2)
```
